specs/HubProtocol.md

```diff
@@ -0,0 +1,225 @@
+* `Result` Message - Indicates results from a previous Invocation message.
+* `Completion` Message - Indicates a previous Invocation has completed, and no further Result messages will be received.
+
+In order to perform a single invocation, Caller follows the following basic flow:
+
+1. Allocate a unique `Invocation ID` value (arbitrary string, chosen by the Caller) to represent the invocation
+2. Send an `Invocation` message containing the `Invocation ID`, the name of the `Target` being invoked, and the `Arguments` to provide to the method.
+3. Wait for a `Result` or `Complete` message with a matching `Invocation ID`
+4. If a `Completion` message arrives, go to 8
+5. If the `Result` message has an error, dispatch the error to the application (i.e. throw an `Exception`, etc.) and complete the invocation immediately.
+6. If the `Result` message has a payload, dispatch the payload to the application (i.e. by yielding a result to an `IObservable`, or by collecting the result for dispatching in step 8)
+7. Go to 3
+8. Complete the invocation, yielding that completion to the Caller (i.e. by completing a `Task` or `IObservable`)
+
+The `Target` of an `Invocation` message must refer to a specific method, overloading is **not** permitted. In the .NET Binder, the `Target` value for a method is defined as the simple name of the Method (i.e. without qualifying type name, since a SignalR endpoint is specific to a single Hub class).
+
+**NOTE**: `Invocation ID`s are arbitrarily chosen by the Caller and the Callee is expected to use the same string in all response messages. Callees may establish reasonable limits on `Invocation ID` lengths and terminate the connection when an `Invocation ID` that is too long is received.
+
+## Multiple Results
+
+The SignalR protocol allows for multiple `Result` messages to be transmitted in response to an `Invocation` message, and allows the receiver to dispatch these results as they arrive, to allow for streaming data from one endpoint to another.
+
+On the Callee side, it is up to the Callee's Binder to determine if a method call will yield multiple results. For example, in .NET, a return type of `IEnumerable` or `IObservable` will yield multiple results, but other return types will yield a single result. The Callee's Binder must encode each result in separate `Result` messages, indicating the end of results by sending a `Completion` message.
+
+On the Caller side, the user code which performs the invocation indicates how it would like to receive the results and it is up the Caller's Binder to determine how to handle the result. If the Caller expects only a single result, but multiple results are returned, the Caller's Binder should yield an error indicating that multiple results were returned. However, if a Caller expects multiple results, but only a single result is returned, the Caller's Binder should yield that single result and indicate there are no further results.
```

Comment:
Just re-read and noticed this was about expecting a single result and receiving multiple results. Since the caller doesn't yield until the `Completion` message arrives, it knows that multiple results were received but only one was expected and can yield the error then. It could even yield the error as soon as the second `Result` message arrives (since it knows it wasn't expecting it).

Receiving an unsolicited `Result` message for an invocation already completed would be a protocol error and we'd terminate the connection.